### DISCUSSION TITLE
Heterogenous GCN Help

### DISCUSSION THREAD

**Discussion by edilaitin** [pyg-team/pytorch_geometric#7314] [Q&A]
Hello everyone,

I am trying to learn GCN and I'm struggling to find example implementations for heterogeneous graphs.

I don't know how to approach the design for my model, I will try to explain the structure of the data and what it should predict.

It is an allocation problem, there are 2 categories of nodes: application components and virtual machines. Both components and virtual machines have node features of different sizes (hardware/software requirements respectively hardware/software specifications).

Between the component nodes there are edges of type 'cc' (component to component) with edge features (as one hot encoding) specifying constraints between the 2 components (eg. these 2 components must not be placed on the same virtual machine).
There are no edges between virtual machines nodes (vm - vm).

Each component has an edge to all virtual machine nodes of type 'linked' or 'unlinked'. The Model should predict for each edge between a component or virtual machine if it is of type 'linked' (this component will be deployed on this virtual machine) or 'unlinked', 

I have multiple training graphs and I want to obtain a model able to predict for a newly given graph all the edge types between component and VMs. (depending only on the node features and component edges).

I have tried so far with RGCN using DGL and HeteroGraphConv({ rel: GraphConv(in_feats, hid_feats) for rel in rel_names}, aggregate='sum') (but I just realized it doesn't even take into account edge features) and a HeteroMLPPredictor but the results are very poor, most of the time it gets stuck predicting only 'unlinked' edges across the whole graph.

Any idea or suggestion (what to read, learn) would help me a lot, thanks!

**Comment by rusty1s** [comment]
For a new graph, do you have knowledge of some linked/unlinked edges? Otherwise, I am not totally sure how a GNN could help here (sure, you can do message passing via edge cc, but there will not be any message exchange between components and VMs). As such, it is probably best to treat this as a two-tower model (GNN on component nodes and MLP on VM nodes), and then take pairs of nodes to do the final binary classification task.

**Reply by edilaitin** [reply]
No, I don't have any knowledge of some linked/unlinked edges for a new graph, does this make it a no-go for GNNs? 

I have tried to follow this implementation from [DGL (Predicting Edge Type of an Existing Edge on a Heterogeneous Graph)](https://docs.dgl.ai/en/0.8.x/guide/training-edge.html#predicting-edge-type-of-an-existing-edge-on-a-heterogeneous-graph) but I can see they have multiple links between their 2 node classes (users and item) not only the ones they are predicting (related to your message exchange point).

Do you know any examples of two-tower models on that can help me? 
Thank you!

**Reply by rusty1s** [reply]
Yeah, as I said. If you don't have links during inference, then you cannot really exchange information between different node types with a GNN. I don't have a paper reference, but the two-tower model would probably look something like
```python
component = gnn(initial_component, edge_index)
vm = mlp(initial_vm)
out = component @ vm.t()
